checker/src/org/checkerframework/checker/upperbound/UpperBoundVisitor.java

```diff
@@ -0,0 +1,131 @@
+        this.IndexFirstArgListMethods.add(
+                TreeUtils.getMethod(
+                        "java.util.List", "get", 1, checker.getProcessingEnvironment()));
+        this.IndexFirstArgListMethods.add(
+                TreeUtils.getMethod(
+                        "java.util.List", "set", 2, checker.getProcessingEnvironment()));
+        //// can't handle until TreeUtils.getMethod has a way to precisly handle method overloading
+        //// this.IndexFirstArgListMethods.add(TreeUtils.getMethod("java.util.List", "remove", 1, checker.getProcessingEnvironment()));
+        this.IndexFirstArgListMethods.add(
+                TreeUtils.getMethod(
+                        "java.util.List", "listIterator", 1, checker.getProcessingEnvironment()));
+        this.IndexFirstArgListMethods.add(
+                TreeUtils.getMethod(
+                        "java.util.List", "addAll", 2, checker.getProcessingEnvironment()));
+        this.IndexFirstArgListMethods.add(
+                TreeUtils.getMethod(
+                        "java.util.List", "add", 2, checker.getProcessingEnvironment()));
+    }
+
+    /**
+     * When the visitor reachs an array access, it needs to check a couple of things. First, it
+     * checks if the index has been assigned a reasonable UpperBound type: only an index with type
+     * LTLengthOf(arr) is safe to access arr. If that fails, it checks if the access is still safe.
+     * To do so, it checks if the MinLen checker knows the minimum length of arr by querying the
+     * MinLenATF. If the MinLen of the array is known, the visitor can check if the index is less
+     * than the MinLen, using the Value Checker. If so then the access is still safe. Otherwise,
+     * report a potential unsafe access.
+     */
+    @Override
+    public Void visitArrayAccess(ArrayAccessTree tree, Void type) {
+        ExpressionTree indexTree = tree.getIndex();
+        ExpressionTree arrTree = tree.getExpression();
+        String arrName = FlowExpressions.internalReprOf(this.atypeFactory, arrTree).toString();
+        AnnotatedTypeMirror indexType = atypeFactory.getAnnotatedType(indexTree);
+
+        // Need to be able to check these as part of the conditional below.
+        // Find max because it's important to determine whether the index is
+        // less than the minimum length of the array. If it could be any
+        // of several values, only the max is of interest.
+        Integer valMax = atypeFactory.valMaxFromExpressionTree(indexTree);
+        Integer minLen = atypeFactory.minLenFromExpressionTree(arrTree);
+
+        // Is indexType LTL/LTOM of a set containing arrName?
+        if ((indexType.hasAnnotation(LTLengthOf.class)
+                        || indexType.hasAnnotation(LTOMLengthOf.class))
+                && (UpperBoundUtils.hasValue(indexType, arrName)
+                        || UpperBoundUtils.hasValue(indexType, arrTree.toString()))) {
```

Comment:
I thought I had commented out localName (mentioned in your comment below) but I had actually commented out lstName. If I comment out both localName and the arrTree.toString(), all tests still pass. So I don't think there's a real problem, and I'm not sure why this code is here. I'll delete it.